OTEP: Extending attributes to support complex values (#4485)
open-telemetry/opentelemetry-specification: oteps/4485-extending-attributes-to-support-complex-values.md

* [Trade-offs and mitigations](#trade-offs-and-mitigations)
  * [Backends don't support `AnyValue` attributes](#backends-dont-support-anyvalue-attributes)
  * [Arbitrary objects are dangerous](#arbitrary-objects-are-dangerous)
* [Future possibilities](#future-possibilities)
  * [Configurable SDK behavior](#configurable-sdk-behavior)
  * [Additional size limits](#additional-size-limits)

## Glossary

In the context of this OTEP, we use the following terminology:

- **Simple attributes** are attributes with primitive types or homogeneous arrays of primitives.
  Their types are known in advance and correspond to the top-level  `string_value`,
  `bool`, `int64`, `double`, and `ArrayValue` of those types in the
  [AnyValue proto definition](https://github.com/open-telemetry/opentelemetry-proto/blob/42319f8b5bf330f7c3dd4a097384f9f6d5467450/opentelemetry/proto/common/v1/common.proto#L28-L40).
  These are currently referred to as *standard* attributes in the
  [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.44.0/specification/common/README.md)

- **Complex attributes** include all other values supported by the `AnyValue` proto,

>   such as maps, heterogeneous arrays, and combinations of those with primitives.

Author: there are a few ways to record empty value as simple or complex attribute - empty string, empty arrayvalue, empty bytes array, empty map. I'd rather avoid mentioning it here.

  via OTTL across all signals, so backends are not free from an obligation to handle
  complex attributes gracefully (even if that just means dropping them currently).
- While backends may still need to add support for complex attributes,
  this is the case with the introduction of any new OTLP feature.
  Bumping the OTLP minor version is already the normal communication mechanism
  for this kind of change.
  A reasonably straightforward implementation option for backends is to just
  JSON serialize complex attributes and store them as strings.

## How

### API

Existing APIs that create or add attributes will be extended to support value
with type `AnyValue`.

It's RECOMMENDED to expose an `AnyValue` API for type checking, ergonomics,
and performance.

Exposing multiple `Attributes` types is NOT RECOMMENDED.

> [!NOTE]
>
> While OTel metrics are designed for aggregation, future SDK implementations
> may choose to record measurements as individual events. This would make complex
> values acceptable in metrics.
>
> Ultimately, it’s up to the backend to decide whether to support complex
> values for a given signal or serialize/drop it otherwise.

> ### SDK

Reviewer: We should add an information that some endpoints (e.g. Prometheus) most likely won't support complex attributes natively (they often do not even support arrays of values). The specification would have to describe how to transform such attributes e.g. in: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/mapping-to-non-otlp.md and some exporters (e.g. Prometheus)  implementation would have to be improved to handle complex attributes.

We should also document in the OTEP how such change affects OpenTracing and OpenCensus bridges.

Author: > We should add an information that some endpoints (e.g. Prometheus) most likely won't support complex attributes natively

addressed in https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/33ed4c2ba4f4b3f84e73aa798eae23d370ec2e01, 

> how such change affects OpenTracing and OpenCensus bridges.

It doesn't, neither of them support complex attributes, so the mapping from them to otel remains exactly the same. And there is no backward mapping, right?

Reviewer: > It doesn't, neither of them support complex attributes, so the mapping from them to otel remains exactly the same. And there is no backward mapping, right?

Right. Yet, the OTEP should describe it explicitly so that it is clear that impact on these bridges has been analyze.

Author: I tried to find a suitable place for "no, it does not affect OC/OT shims" and couldn't. There are a lot of things this  proposal doesn't affect and listing them all would be hard :) 

So I'd prefer not to mention this.

**The OTel SDK MUST support complex attributes on all telemetry signals.**

It includes spans, logs, and profiles where complex attributes have known use-cases as
well as metrics, resources, instrumentation scope, and entities where complex attributes
usage is currently discouraged and expected to be accidental.

Any future OTel signals SHOULD support complex attributes unless there is a
reason not to, which MUST be documented in the specification.

The OTel SDK SHOULD pass both simple and complex attributes to the processing
pipeline as `AnyValue`.
The SDK MUST support reading and modifying complex attributes during processing.

#### Attribute limits

The SDK SHOULD apply [attribute limits](/specification/common/README.md#attribute-limits)
to complex attributes.

**Attribute value length limit** SHOULD be applied to all leaf string nodes in
`AnyValue` which SHOULD be truncated to the configured limit.

Leaf nodes of an `AnyValue` attribute SHOULD count toward the **attribute count limit**.
If the limit is reached, the SDK MUST drop the entire `AnyValue` attribute;

> partial exports are not allowed.

Author: created https://github.com/open-telemetry/opentelemetry-specification/issues/4487, added a comment to the OTEP.

### Semantic conventions

Semantic conventions will be updated with the following guidance:

- Simple attributes SHOULD be used whenever possible. Semantic conventions SHOULD
  assume that backends do not index individual properties of complex attributes,
  that querying or aggregating on such properties is inefficient and complicated,
  and that reporting complex attributes carries higher performance overhead.

- Complex attributes that are likely to be large (when serialized to a string) SHOULD
  be added as *opt-in* attributes. *Large* will be defined based on common backend
  attribute limits, typically around 4–16 KB.

- Complex attributes MUST NOT be used on metrics, resources, instrumentation scopes,
  or as identifying attributes on entities.

## Trade-offs and mitigations

### Backends don't support `AnyValue` attributes

While it should be possible to record complex data in telemetry, many backends do not
support it, which can result in individual complex attributes being dropped.

> We mitigate this through:

Reviewer: We can also add some experimental environmental variable that controls the OTLP exporters to flatten (JSONify) the complex attributes (something similar to `OTEL_SEMCONV_STABILITY_OPT_IN`). Having this as env var would make it possible to drop this feature after we give enough time for the backends to add support for complex attributes.

Author: I think Trask addressed it in the https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/8a219900aaba1647c65157a2276e0e16b170ab93 - we can provide the option in exporter, but at least initially try to lean on the collector.

  the impact of unsupported attribute types to early adopters, while giving
  backends time to add support.

- Semantic conventions guidance that limits usage of complex attributes.

- Existing collector transformation processor that can drop, flatten, serialize,
  or truncate complex attributes using OTTL.

### Arbitrary objects are dangerous

Allowing arbitrary objects as attributes is convenient but increases the risk of
including large, sensitive, mutable, non-serializable, or otherwise problematic
data in telemetry.

It is RECOMMENDED to enforce `AnyValue` compatibility at the API level.
Users and instrumentations SHOULD define custom `AnyValue`-compatible models
to minimize misuse and reduce performance overhead.

OTel SDKs MAY provide convenience methods to convert arbitrary objects to `AnyValue`.
If such convenience is provided, it is RECOMMENDED to limit supported types
to primitives, arrays, standard library collections, named tuples, JSON objects, and
similar structures following [mapping to OTLP AnyValue](/specification/common/attribute-type-mapping.md#mapping-arbitrary-data-to-otlp-anyvalue).
Falling back to a string representation of unknown objects is RECOMMENDED to

> minimize the risk of unintentional use of complex attributes.

Reviewer: The Go log bridges we have such function
https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/bridges/otelslog/convert.go.
However, it is not publicly available in order to minimize misuse and reduce performance overhead for direct users Logs API. People can "copy-paste" it, but then it is "their" problem.

If you want you can give this as an example.

Author: It's already part of the spec and it's there to support log bridges. Plus there are SDKs like Go [or .NET](https://github.com/open-telemetry/opentelemetry-dotnet/blob/71abd4169b4b6c672343b37c32e3337bc227ed32/src/OpenTelemetry/Logs/ILogger/OpenTelemetryLogger.cs#L134) which do at least some of it already. 

The point of this section is to limit the support for arbitrary objects and address concerns expressed previously around someone throwing random things in.

I rephrased it to 

> OTel SDKs that provide convenience to convert arbitrary objects to `AnyValue` SHOULD limit supported types...

LMK if it better conveys 'be careful with this' message.

Reviewer: The message is OK. I just thought that maybe you want to have/give some examples :wink: 
Feel free to resolve.

Author: thanks for the explanation, added some links in https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/017b9fdd749ff8f9487d0f86599cd4139cfb3dc2

## Future possibilities

### Configurable SDK behavior

The SDK behavior for complex attributes can be made customizable on a per-signal

> basis, allowing it to either **serialize complex values to JSON strings** or

Reviewer: Do we want to drop/JSONify early or at the end?
Should we call out the possible places where the serialization complex values to JSON strings happens?
If we accept JSONifying at the end then maybe we can add "or per-exporter" basis?

Author: Agree, I think https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/8a219900aaba1647c65157a2276e0e16b170ab93 addresses it

> ### Configurable SDK behavior
> 
> The SDK behavior for complex attributes can be made customizable on a per-signal
> basis, allowing it to either **serialize complex values to JSON strings** or
> **drop** the corresponding attribute.

Reviewer 2: related to @pellared's https://github.com/open-telemetry/opentelemetry-specification/pull/4485/files#r2055395613

```suggestion
### Configurable OTLP exporter behavior

The OTLP exporter behavior for complex attributes can be made customizable on a per-signal
basis, allowing complex attributes to be:
- passed through (the default),
- serialized to JSON, or
- dropped
```

[opentelemetry-java#7123](https://github.com/open-telemetry/opentelemetry-java/pull/7123) and
[opentelemetry-go#6180](https://github.com/open-telemetry/opentelemetry-go/pull/6180)),
extending the standard attributes provides a more seamless and user-friendly API experience.

### Why isn't it a breaking change?

Currently, the SDK specification has a clause that says extending
the set of standard attribute would be
[considered a breaking change](/specification/common/README.md#standard-attribute).

We believe that removing this clause and extending standard
attributes will not in fact be a breaking change for
anyone in the OpenTelemetry ecosystem:

- Language SDKs can implement this without breaking their backwards
  compatibility guarantees (e.g., [Java's](https://github.com/open-telemetry/opentelemetry-java/blob/main/VERSIONING.md)).
- The OpenTelemetry Collector already supports transforming data to complex attributes
  via OTTL across all signals, so backends are not free from an obligation to handle
  complex attributes gracefully (even if that just means dropping them currently).
- While backends may still need to add support for complex attributes,

>   this is the case with the introduction of any new OTLP feature.
>   Bumping the OTLP minor version is already the normal communication mechanism
>   for this kind of change.

Reviewer: What OTLP do you have in mind? OTLP exporter? OTLP proto? I guess that the exporter as the proto already "supports" it.

```suggestion
  this is the case with the introduction of any new OTLP exporter feature.
  Bumping the OTLP exporter minor version is already the normal communication mechanism
  for this kind of change.
```

  A reasonably straightforward implementation option for backends is to just
  JSON serialize complex attributes and store them as strings.

## How

### API

Existing APIs that create or add attributes will be extended to support value
with type `AnyValue`.

It's RECOMMENDED to expose an `AnyValue` API for type checking, ergonomics,
and performance.

Exposing multiple `Attributes` types is NOT RECOMMENDED.

> [!NOTE]
>
> While OTel metrics are designed for aggregation, future SDK implementations
> may choose to record measurements as individual events. This would make complex

> > values acceptable in metrics.

Reviewer 3: I don't buy this logic. A possible future event-based metrics SDK does not remove the need to consider client-side aggregation scenarios. If complex attributes are not acceptable for local aggregation, they don't become acceptable because they're not a problem for a different feature.

Reviewer 3: The motivation as I understand it is to have just a single attribute type. Introducing a new type for Attributes would be breaking for some languages in existing APIs where extending the existing type is not. Because that type is used in multiple places, extending it in one place extends it in all places.

I should point out that I don't love that argument. It's just the one that's been explained to me.

NB I am not a maintainer of any such language. JS (which I do maintain) would allow for it.

Author: > A possible future event-based metrics SDK does not remove the need to consider client-side aggregation scenarios. 

It is possible and could even be useful to have a metric implementation that records event for each measurement instead of  aggregating them in memory and exporting an aggregation.

> If we don't see any benefit for Metrics, then why make Metrics also support complex attributes?

@dyladan's absolutely right that the key motivation is to avoid having multiple attribute types. It's for ergonomics reasons mentioned in the why section here, specifically https://github.com/open-telemetry/opentelemetry-specification/pull/4201. 

@cijothomas you probably noticed that having different collections for .NET tracing and metrics attributes results in a lot of boilerplate at the instrumentation time. It also results in inconsistencies (Java calls them `ExtendedAttributes`, Python calls them `LogAttributes`, and attempt to just document the name raises even more questions - https://github.com/open-telemetry/opentelemetry-specification/pull/4462)

Reviewer 2: @dyladan Is your performance concern addressed by the new OTEP language?

> Simple attributes SHOULD be used whenever possible. Instrumentations SHOULD assume that backends do not index individual properties of complex attributes, that querying or aggregating on such properties is inefficient and complicated, and that reporting complex attributes carries higher performance overhead.

Author: I removed the note

>
> Ultimately, it’s up to the backend to decide whether to support complex
> values for a given signal or serialize/drop it otherwise.

### SDK

> **The OTel SDK MUST support complex attributes on all telemetry signals.**

Reviewer 3: Biggest question I have is how the SDK is expected to aggregate metrics with complex attribute values. Are we expected to calculate deep object equality? It will make it significantly more expensive to calculate metric identity for aggregation and make the code significantly more complex, but semconv will never define a metric that actually uses it.

Reviewer 2: > It will make it significantly more expensive to calculate metric identity for aggregation

yeah, definitely. the OTEP includes this language: "Semantic conventions SHOULD assume that backends do not index individual properties of complex attributes, that querying or aggregating on such properties is inefficient and complicated, and that reporting complex attributes carries higher performance overhead."

I think it makes sense to add this warning in the spec itself to make sure that the warning is carried over into API docs

> and make the code significantly more complex

A quick look at the Java SDK looks like the existing metric aggregation code shouldn't need to be updated since we already have (deep) identity defined for "anyvalue" which then rolls up to the existing identity defined for the "attributes" set, but copying @jack-berg to keep me honest.

Author: > I think it makes sense to add this warning in the spec itself to make sure that the warning is carried over into API docs

I added a section on this in the https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/017b9fdd749ff8f9487d0f86599cd4139cfb3dc2

> It will make it significantly more expensive to calculate metric identity for aggregation and make the code significantly more complex, but semconv will never define a metric that actually uses it.

also added a note on AnyValue implementation covering hash code and equality. Keep in mind that using complex attributes on metrics (resources, etc) is NOT RECOMMENDED, so it won't be a mainstream case. 
It's hard to imagine that someone intentionally puts deeply nested object into metric and having a tuple would have the same perf implications as a short array (yet arrays are allowed)

It's much more likely that someone puts hard-cardinality attribute than deeply nested one. Yet, we don't block arbitrary string usage.

Having said that, it's possible to make it reasonably efficient if `AnyValue` hashcode is cached and re-calculated on write. 
It'd be good enough for resource/instr scope/entity attributes which can easily cache `AnyValue` if they happen to use it.

Reviewer 4: > A quick look at the Java SDK looks like the existing metric aggregation code shouldn't need to be updated since we already have (deep) identity defined for "anyvalue" which then rolls up to the existing identity defined for the "attributes" set, but copying @jack-berg to keep me honest.

Yeah I agree that it would be more expensive to compute equality (though maybe not different than computing equality for a large string array). Code complexity shouldn't be an issue because as you alude to, you it ends up being a recursive equality implementation, with little code change needed to accommodate the extra structure. Of course a recursive implementation opens up the possibility of stack overflow attacks, and so we may want to think about some sort of max depth attribute limit. But my understanding is that all (most?) recursive algos can be translated to a sequential algorithm, so maybe not a problem.

Reviewer 2: Resolving as I think we've addressed @dyladan's concern here by loosening the SDK requirement to

> OTel SDK MUST support setting complex attributes on spans, logs, profiles,
descriptive entity attributes.
>
> OTel SDK MAY support setting complex attributes on other telemetry signals.

Please open a new discussion and summarize if there are still open issues here, thanks!

It includes spans, logs, and profiles where complex attributes have known use-cases as

> well as metrics, resources, instrumentation scope, and entities where complex attributes

Reviewer 3: >This is not just a client aggregation issue either. Even if we decide it's ok to use simple object equality and let backends sort it out, it needs to be handled somewhere. OTLP is defined such that this is disallowed, so backends may be dropping, serializing, or doing something else entirely with these values.

I believe this will be a prometheus compatibility issue at least

> well as metrics, resources, instrumentation scope, and entities where complex attributes

Reviewer: I agree that to have proper behavior then it would have to be deep equality.

> well as metrics, resources, instrumentation scope, and entities where complex attributes

Author: I think it's the same discussion as https://github.com/open-telemetry/opentelemetry-specification/pull/4485#discussion_r2059380596.

AnyValue need efficient hash code and deep equality check - it needs it already since metrics support array attributes.
Given it's not recommended to add complex attributes to metrics (and seem very weird to add deeply nested complex objects as metric attributes), the whole scenario should not be a mainstream one. I.e. the perf impact of someone accidentally adding a complex object to metrics would be limited to this app and won't affect the mainstream case.

Prometheus already serializes everything [non-string to JSON string (including arrays)](https://github.com/open-telemetry/opentelemetry-specification/blob/74c626723da290127a1d9d6bb18f9117235e7f10/specification/compatibility/prometheus_and_openmetrics.md#metric-attributes), which is already inefficient but also rarely (if ever) used. 

Once entities are in, only identifying entity attributes would be part of the metric identity, so there won't be a problem with legacy (at that point) resources. I can mention it in the OTEP language if you feel it's an important detail.

> well as metrics, resources, instrumentation scope, and entities where complex attributes

Reviewer 2: Resolving as I think we've addressed @dyladan's concern here by loosening the SDK requirement to

> OTel SDK MUST support setting complex attributes on spans, logs, profiles,
descriptive entity attributes.
>
> OTel SDK MAY support setting complex attributes on other telemetry signals.

Please open a new discussion and summarize if there are still open issues here, thanks!

usage is currently discouraged and expected to be accidental.

Any future OTel signals SHOULD support complex attributes unless there is a
reason not to, which MUST be documented in the specification.

The OTel SDK SHOULD pass both simple and complex attributes to the processing
pipeline as `AnyValue`.
The SDK MUST support reading and modifying complex attributes during processing.

#### Attribute limits

The SDK SHOULD apply [attribute limits](/specification/common/README.md#attribute-limits)
to complex attributes.

**Attribute value length limit** SHOULD be applied to all leaf string nodes in
`AnyValue` which SHOULD be truncated to the configured limit.

Leaf nodes of an `AnyValue` attribute SHOULD count toward the **attribute count limit**.
If the limit is reached, the SDK MUST drop the entire `AnyValue` attribute;
partial exports are not allowed.

> <!-- See https://github.com/open-telemetry/opentelemetry-specification/issues/4487,
> we should define limits even if this OTEP is not merged -->

Reviewer: ```suggestion
> [!NOTE]
>
> See [#4487](https://github.com/open-telemetry/opentelemetry-specification/issues/4487),
> we should define limits even if this OTEP is not merged.
```

### Exporters

OTLP exporter SHOULD, by default, pass `AnyValue` attributes to the endpoint.

Exporters for protocols that do not natively support complex values, such as Prometheus,
SHOULD represent complex values as JSON-encoded strings following

> [attribute specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute).

Reviewer: ```suggestion
[attribute specification](/specification/common/README.md#attribute).
```

  or truncate complex attributes using OTTL.

### Arbitrary objects are dangerous

Allowing arbitrary objects as attributes is convenient but increases the risk of
including large, sensitive, mutable, non-serializable, or otherwise problematic
data in telemetry.

It is RECOMMENDED to enforce `AnyValue` compatibility at the API level.
Users and instrumentations SHOULD define custom `AnyValue`-compatible models
to minimize misuse and reduce performance overhead.

OTel SDKs that provide convenience to convert arbitrary objects to `AnyValue`
SHOULD limit supported types to primitives, arrays, standard library collections,
named tuples, JSON objects, and similar structures following
[mapping to OTLP AnyValue](/specification/common/attribute-type-mapping.md#mapping-arbitrary-data-to-otlp-anyvalue).

Falling back to a string representation of unknown objects is RECOMMENDED to
minimize the risk of unintentional use of complex attributes.

## Future possibilities

### Configurable OTLP exporter behavior

The OTLP exporter behavior for complex attributes can be made customizable on a per-signal
basis, allowing complex attributes to be:

> - passed through (the default),

Reviewer: to fix lint error

```suggestion

- passed through (the default),
```

  [spans and events are often thought of as being conceptual
  twins](https://opentelemetry.io/blog/2025/opentelemetry-logging-and-you/#how-is-this-different-from-other-signals),
  and that the choice between modeling something as a span or an event should not
  be influenced by whether complex attributes are needed.

### Why do we want to extend standard attributes?

Instead of introducing a second set of "extended" attributes that can be used on
spans and events, we propose to extend the standard attributes.

Having multiple attribute sets across the API
[creates ergonomic challenges](https://github.com/open-telemetry/opentelemetry-specification/issues/4201).
While there are some mitigations (as demonstrated in
[opentelemetry-java#7123](https://github.com/open-telemetry/opentelemetry-java/pull/7123) and
[opentelemetry-go#6180](https://github.com/open-telemetry/opentelemetry-go/pull/6180)),
extending the standard attributes provides a more seamless and user-friendly API experience.

### Why isn't it a breaking change?

Currently, the SDK specification has a clause that says extending
the set of standard attribute would be

> [considered a breaking change](/specification/common/README.md#standard-attribute).

Reviewer 4: I mentioned this example in last week's spec meeting:

Based on the current spec language stating that standard attributes are primitives and arrays of primitives, and that extending the standard attribute definition is a breaking change, it's conceivable that a OTLP receiver would receive this and reject data an entire record when it encounters attributes which don't conform to this definition. Of course given the proto definition allowing a more expansive definition, you can argue that a receiver would be better off taking a lenient stance and simply skipping these attributes, but we don't define the expected behavior of OTLP receivers with data which is allowable at the proto level, but semantically invalid according to the specification.

An OTLP receiver that behaves like this would begin rejecting spans that were previously accepted once instrumentation is updated to encode complex as AnyValue attributes.

Reviewer 2: @jack-berg just confirming you remain unconvinced by this part of the OTEP:

> We believe that removing this clause and extending standard attributes will not in fact be a breaking change within the OpenTelemetry ecosystem:
> - ...
> - The OpenTelemetry Collector already supports transforming data to complex attributes via OTTL across all signals, so backends are not free from an obligation to handle complex attributes gracefully (even if that just means dropping them currently).

what do you think about this additional reason (not included so far):

- The clause about extending standard attributes was never applied to the OTLP protocol, and so OTLP receivers were not given this guarantee.

Reviewer 4: > so backends are not free from an obligation to handle complex attributes gracefully (even if that just means dropping them currently)

I haven't seen that obligation play out in practice with my employer.

Full disclosure - my employer's OTLP implementation currently rejects spans with any value attributes. We've been working to change that independently of this discussion, and our change is scheduled to roll out before any changes from this OTEP could take affect (i.e. I don't think we should block b/c of my employer). This implementation decision was made years ago by folks other than myself and before the spec had the language prohibiting extending standard attributes. For me, its a data point illustrating what kinds of assumptions engineers make and how things may be breaking changes in unintuitive ways. Its impossible to know of all the OTLP receivers out there and their behavior, and so the key question for me is there a reasonable probability that other OTLP receivers took the spec's language seriously and implemented strict validation. I know of at least one case so I think its possible there are others.  

As I mentioned in chat last week's spec SIG, I think that having a default SDK behavior to JSON serialize any values is the price we have to pay for making a mistake. Similar to how we are stuck with bad explicit histogram default boundaries, and with explicit histograms as the default implementation instead of exponential histograms.

Author: I believe there is a big difference with explicit histogram boundaries - it is possible to report histograms today and there are plenty of existing telemetry producers and consumers that have expectations. Changing the default in the SDK would change the telemetry produced by all of them.

Here, we're talking about trying to preserve a behavior for something that doesn't exist yet. Nobody reports complex attributes on spans today and nothing will be broken until someone starts producing them. The producers themselves will be experimental for as long as we feel necessary (we don't have to stabilize the API until we feel comfortable).

Reviewer 5: Out of curiosity, @jack-berg you think that providing a new option for opt-in serialization to JSON strings would be not enough? (given that the default would be sending such AnyValue items instead of JSONing them)

Reviewer 2: > my employer's OTLP implementation currently rejects spans with any value attributes

I think this goes to @jsuereth's comment in last week's SIG that we should bump the proto (minor) version if we go forward, since proto version is the standard signal for backends to follow and implement new (or in this case existing but mostly unused) proto fields.

And then we can say that SDKs can only emit complex attributes using that proto version or later.

And if we think that backends need time to adapt, we can require a minimum grace period before stable SDKs emit complex attributes, similar to what we did with HTTP semconv when we included "stable HTTP semconv SHOULD NOT be released prior to October 1, 2023".

Reviewer 2: based on this discussion, I refined the reasons given in the OTEP why I think this is non-breaking.

I removed

> - The OpenTelemetry Collector already supports transforming data to complex attributes via OTTL across all signals, so backends are not free from an obligation to handle complex attributes gracefully (even if that just means dropping them currently).

since I think @jack-berg debunked that one pretty well.

and under "While backends may still need to add support for complex attributes, this is the case with the introduction of any new OTLP feature", I added these additional bullets:

  - SDKs will be required to only emit complex attributes under that OTLP version or later.
  - If we get pushback that backends need more time to adapt to the new OTLP minor version, we can add a grace period similar to the one we had for HTTP semconv when we included: "Stable HTTP semconv SHOULD NOT be released prior to October 1, 2023."

Reviewer 4: This convo came up in the 4/29/25 spec call. Going to try to reiterate some of the points made there.

> Here, we're talking about trying to preserve a behavior for something that doesn't exist yet. Nobody reports complex attributes on spans today and nothing will be broken until someone starts producing them.

I think that what you're saying is that because no instrumentation produces these today, nothing can be broken until we: 1. add complex attribute values 2. users update to new versions of instrumentation which produce them. While its technically true that nothing breaks by adding support for new attribute values, in practice, we'll likely see users that fail to see the difference between API / SDK adding support for something and instrumentation updating to produce it. In practice, they'll update a bunch of versions together (both API, SDK, and instrumentation) and the collective stack will start producing data which a backend will reject if it took the advice in the specification about standard attributes seriously and implemented strict validation.

> Out of curiosity, @jack-berg you think that providing a new option for opt-in serialization to JSON strings would be not enough? (given that the default would be sending such AnyValue items instead of JSONing them)

My argument is all about the default. I think its obvious that we should provide SDK options to allow a user to emit JSON stringified versions of complex attributes, or the structured version. The default matters because emiting the structured version by default is appealing in the long term. I.e. 10 years from now we could be far past the point of any OTLP receiver not supporting complex values, at which point having SDKs produce JSON stringified versions will seem silly. And yet it will also be difficult to change that default without a major version bump, which I personally want to avoid. 

I thought @trask brough up a good idea in the spec SIG. It was something to the effect of: 

- Add support for complex attribute values at API / SDK level
- Provide options to emit either the structured of JSON stringified representation, **but initially emit _neither_ by default**
- Make an announcement that at some point TBD date in the future, SDKs will start emitting the structured version

This seems to strike a nice balance. It acknowledges that this _probably_ isn't a breaking change at the proto level, but that it is a very significant change, especially given the strict language in the specification regarding standard attribute values. It allows us to land where we want in the long term (i.e. SDKs default structured representation by default) without SDKs needing to be in the tough position of emitting JSON stringified by default today and later having to change that.

One question I have: what happens with metric identity if a user doesn't want complex attributes emitted at all? It seems like the SDK would need to strip complex attributes before aggregation, since if they are only stripped at export time (e.g in the OTLP exporter), we'll emit metrics with duplicate identities.

Author: thanks @jack-berg for the detailed explanation!

To address the metric question

> One question I have: what happens with metric identity if a user doesn't want complex attributes emitted at all? It seems like the SDK would need to strip complex attributes before aggregation, since if they are only stripped at export time (e.g in the OTLP exporter), we'll emit metrics with duplicate identities.

Given that we discourage complex attributes use on metrics and would not have them in semantic conventions/OTel instrumentations, I would expect them to be used by accident. So, dropping them in view and/or changing the producer to stop emitting them is a possible way.

Still, if complex attributes are not too big and too complex, they won't be a problem. Prometheus supports them with remote-write (serialized to string).

The TL;DR: I believe we don't need a config to drop all complex attributes on all metrics, having individual attribute disablement via views should be sufficient.

Reviewer 2: I think we have tentative agreement based on spec SIG meeting discussions, and I'm resolving this based on this language that is now added to the OTEP:

> Stable SDKs will be prohibited from emitting complex attributes on signals other than Logs until at least Oct 1, 2025.

> that querying or aggregating on such properties is inefficient and complicated,
> and that reporting complex attributes carries higher performance overhead.

> [!NOTE]
>
> While OTel metrics are designed for aggregation, future SDK implementations
> may choose to record measurements as individual events. This would make complex
> values acceptable in metrics.
>
> Ultimately, it’s up to the backend to decide whether to support complex
> values for a given signal or serialize/drop it otherwise.

### SDK

**The OTel SDK MUST support complex attributes on all telemetry signals.**

It includes spans, logs, and profiles where complex attributes have known use-cases as
well as metrics, resources, instrumentation scope, and entities where complex attributes
usage is currently discouraged and expected to be accidental.

Any future OTel signals SHOULD support complex attributes unless there is a
reason not to, which MUST be documented in the specification.

The OTel SDK SHOULD pass both simple and complex attributes to the processing
pipeline as `AnyValue`.
The SDK MUST support reading and modifying complex attributes during processing.

#### `AnyValue` implementation notes

> `AnyValue` implementation SHOULD provide efficient hash code and equality check

Author: ```suggestion
`AnyValue` implementation SHOULD provide efficient hash code and deep equality check
```

since its possible (but not recommended) to use complex attributes as a resource,
instrumentation scope, or metric attribute where hash code and equality may be
extensively used to identify tracer/meter/logger or the time series.

#### Attribute limits

The SDK SHOULD apply [attribute limits](/specification/common/README.md#attribute-limits)
to complex attributes.

**Attribute value length limit** SHOULD be applied to all leaf string nodes in
`AnyValue` which SHOULD be truncated to the configured limit.

Leaf nodes of an `AnyValue` attribute SHOULD count toward the **attribute count limit**.
If the limit is reached, the SDK MUST drop the entire `AnyValue` attribute;
partial exports are not allowed.

> [!NOTE]
>
> We should define `AnyValue` limits even if this OTEP is not merged -
> [#4487](https://github.com/open-telemetry/opentelemetry-specification/issues/4487)

### Exporters

OTLP exporter SHOULD, by default, pass `AnyValue` attributes to the endpoint.

Exporters for protocols that do not natively support complex values, such as Prometheus,
SHOULD represent complex values as JSON-encoded strings following

> [attribute specification](/specification/common/README.md#attribute).

Reviewer 3: I think we will need some sort of stable JSON stringify right? 

`{ "a": "b", "c": "d" }` should be equivalent to `{ "c": d", "a": "b" }`

Reviewer: FWIW Go's [`json.Marshal`](https://pkg.go.dev/encoding/json#Marshal) sorts maps:

> The map keys are sorted and used as JSON object keys

Author: great point, added a note in https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/95f2bb6faf7c49857d383987b89c3819d56122c7, there might be some other settings to document (like null/empty handling), but this list would be quite short since we effectively serialize existing json object to string and most of the edge cases were handled by conversion to anyvalue and exporter.

> may choose to record measurements as individual events. This would make complex
> values acceptable in metrics.
>
> Ultimately, it’s up to the backend to decide whether to support complex
> values for a given signal or serialize/drop it otherwise.

### SDK

**The OTel SDK MUST support complex attributes on all telemetry signals.**

It includes spans, logs, and profiles where complex attributes have known use-cases as
well as metrics, resources, instrumentation scope, and entities where complex attributes
usage is currently discouraged and expected to be accidental.

Any future OTel signals SHOULD support complex attributes unless there is a
reason not to, which MUST be documented in the specification.

The OTel SDK SHOULD pass both simple and complex attributes to the processing
pipeline as `AnyValue`.
The SDK MUST support reading and modifying complex attributes during processing.

#### `AnyValue` implementation notes

`AnyValue` implementation SHOULD provide efficient hash code and deep equality check
since its possible (but not recommended) to use complex attributes as a resource,
instrumentation scope, or metric attribute where hash code and equality may be

> extensively used to identify tracer/meter/logger or the time series.

Reviewer 2: Similar to https://github.com/open-telemetry/opentelemetry-specification/issues/4443.

(and I noticed that the Java's AnyValue implementation doesn't do this currently)

```suggestion
extensively used to identify tracer/meter/logger or the time series.

Equality of `KeyValueList`s MUST be unaffected by the ordering of their `KeyValue` pairs.
```

Having multiple attribute sets across the API
[creates ergonomic challenges](https://github.com/open-telemetry/opentelemetry-specification/issues/4201).
While there are some mitigations (as demonstrated in
[opentelemetry-java#7123](https://github.com/open-telemetry/opentelemetry-java/pull/7123) and
[opentelemetry-go#6180](https://github.com/open-telemetry/opentelemetry-go/pull/6180)),
extending the standard attributes provides a more seamless and user-friendly API experience.

### Why isn't it a breaking change?

Currently, the SDK specification has a clause that says extending
the set of standard attribute would be
[considered a breaking change](/specification/common/README.md#standard-attribute).

We believe that removing this clause and extending standard
attributes will not in fact be a breaking change within the OpenTelemetry ecosystem:

- Language SDKs can implement this without breaking their backwards
  compatibility guarantees (e.g., [Java's](https://github.com/open-telemetry/opentelemetry-java/blob/main/VERSIONING.md)).
- While backends may still need to add support for complex attributes,
  this is the case with the introduction of any new OTLP feature.

>   - Bumping the OTLP minor version is already the normal communication mechanism

Author: added Proto section

  * [Arbitrary objects are dangerous](#arbitrary-objects-are-dangerous)
* [Prototypes](#prototypes)
* [Future possibilities](#future-possibilities)
  * [Configurable OTLP exporter behavior](#configurable-otlp-exporter-behavior)
  * [Additional size limits](#additional-size-limits)
* [Backend research](#backend-research)

## Glossary

In the context of this OTEP, we use the following terminology:

- **Simple attributes** are attributes with primitive types or homogeneous arrays of primitives.
  Their types are known in advance and correspond to the top-level  `string_value`,
  `bool`, `int64`, `double`, and `ArrayValue` of those types in the
  [AnyValue proto definition](https://github.com/open-telemetry/opentelemetry-proto/blob/42319f8b5bf330f7c3dd4a097384f9f6d5467450/opentelemetry/proto/common/v1/common.proto#L28-L40).
  These are currently referred to as *standard* attributes in the
  [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.44.0/specification/common/README.md)

- **Complex attributes** include all other values supported by the `AnyValue` proto,

>   such as null (empty) value, maps, heterogeneous arrays, and combinations of those with primitives.

Reviewer 6: Is supporting JSON the reason for adding an empty value?

Reviewer: I think that the reason is also the parity with https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#type-any

Reviewer 6: Found a related PR https://github.com/open-telemetry/opentelemetry-specification/pull/3853 😉

Reviewer: Also in OTLP since Jun 9, 2020: https://github.com/open-telemetry/opentelemetry-proto/pull/157

Reviewer 2: I think we can resolve this, null/empty value is part of

> all other values supported by the `AnyValue` proto

### Why do we want to extend standard attributes?

Instead of introducing a second set of "extended" attributes that can be used on
spans and events, we propose to extend the standard attributes.

Having multiple attribute sets across the API
[creates ergonomic challenges](https://github.com/open-telemetry/opentelemetry-specification/issues/4201).
While there are some mitigations (as demonstrated in
[opentelemetry-java#7123](https://github.com/open-telemetry/opentelemetry-java/pull/7123) and
[opentelemetry-go#6180](https://github.com/open-telemetry/opentelemetry-go/pull/6180)),
extending the standard attributes provides a more seamless and user-friendly API experience.

### Why doesn't this require a major version bump?

Currently, the SDK specification has a clause that says extending
the set of standard attribute would be
[considered a breaking change](/specification/common/README.md#standard-attribute).

We believe that removing this clause and extending standard
attributes can be done gracefully across the OpenTelemetry ecosystem
without requiring a major version bump:

- Language SDKs can implement this without breaking their backwards

>   compatibility guarantees (e.g., [Java's](https://github.com/open-telemetry/opentelemetry-java/blob/main/VERSIONING.md)).

Author: thanks! added link to prototypes section

- While backends may still need to add support for complex attributes,
  this is the case with the introduction of any new OTLP feature.
  - Bumping the OTLP minor version is already the normal communication mechanism
    for this kind of change.
  - SDKs will be required to only emit complex attributes under that OTLP version

>     or later.

Reviewer: I do not understand this statement. Is there anything in OTLP payload that says about the version?

Reviewer 2: resolving, we discuss in Log SIG meeting

>   - Stable SDKs and exporters will be prohibited from emitting complex attributes by default on signals

Reviewer: I think this would be more clear and precise

```suggestion
  - Stable exporters will be prohibited from emitting complex attributes by default on signals
```

Especially given

>  The SDK MUST support reading and modifying complex attributes during processing.

>     other than Logs until at least Oct 1, 2025.

Reviewer: Cannot a stable exporters provide an opt-in development env variable that enables emitting complex attributes on all signals?

Notice that the OTEP already says:

> The SDK MUST support reading and modifying complex attributes during processing.

Author: ```suggestion
  - By default, stable SDKs will be prohibited from emitting complex attributes on signals
    other than Logs until at least Oct 1, 2025.
```

it should be ok to do it as opt-in similarly to semconv?

  - A reasonably straightforward implementation option for backends is to just
    JSON serialize complex attributes and store them as strings.

## How

### API

Existing APIs that create or add attributes will be extended to support value
with type `AnyValue`.

It's RECOMMENDED to expose an `AnyValue` API for type checking, ergonomics,
and performance.

> Exposing multiple `Attributes` types is NOT RECOMMENDED.

Reviewer 2: ```suggestion
Exposing multiple types of attribute sets is NOT RECOMMENDED, such as having "ExtendedAttributes" in addition "Attributes".
```

OTel API MUST support setting complex attributes on spans, logs, profiles,

> span links, exemplars, and as descriptive entity attributes.

Reviewer 2: ```suggestion
span links, and as descriptive entity attributes.
```

> span links, exemplars, and as descriptive entity attributes.

Reviewer 2: oops, thanks! I forgot to hit "commit suggestion"

> span links, exemplars, and as descriptive entity attributes.

Reviewer 7: Should we also allow for identifying entity attributes?

Reviewer: Why not, right?

Reviewer 7: I don't see a reason to single out entity identifying attributes. IMO, let's aim for uniformness: every attribute should be an extended attribute unless there are good reasons to restrict it.

OTel API MAY support setting complex attributes on metrics, resources,
and instrumentation scope. OTel API SHOULD NOT provide convenience methods
to set complex attributes on these signals. APIs that allow to set complex attributes
on these signals (for example, providing an arbitrary attributes collection when

> recording a metric), MUST document complex attributes usage limitations.

Author: let's document limitations for all APIs, not just metrics/resources

> Simple attributes SHOULD be used whenever possible. Instrumentations SHOULD
> assume that backends do not index individual properties of complex attributes,
> that querying or aggregating on such properties is inefficient and complicated,
> and that reporting complex attributes carries higher performance overhead.

### SDK

OTel SDK MUST support setting complex attributes on spans, logs, profiles,
span links, exemplars, and as descriptive entity attributes.

> OTel API MAY support setting complex attributes on metrics, resources,

Reviewer 5: OTel SDK?

instrumentation scope, and span events.

The SDK MUST support reading and modifying complex attributes during processing.

#### `AnyValue` implementation notes

> `AnyValue` implementation SHOULD provide efficient hash code and deep equality check

Reviewer 5: > Having that said, I think that the current description is good enough for an OTEP (and the current state of the existing implementations). We can think how to make the description better when it will be part of the the actual specification.

Same feeling, the current description is enough for this OTEP. Some of this notes can be included as alternatives/options which can be discussed when this content lands in the Spec, etc.

> `AnyValue` implementation SHOULD provide efficient hash code and deep equality check
> since its possible (but not recommended) to use complex attributes as a resource,
> instrumentation scope, or metric attribute where hash code and equality may be
> extensively used to identify tracer/meter/logger or the time series.

Reviewer 2: ```suggestion
`AnyValue` implementation SHOULD provide deep equality check
since its possible (but not recommended) to use complex attributes as a resource,
instrumentation scope, or metric attribute where equality may be
extensively used to identify tracer/meter/logger or the time series.
```

Equality of `KeyValueList`s MUST be unaffected by the ordering of their `KeyValue` pairs.

#### Attribute limits

The SDK SHOULD apply [attribute limits](/specification/common/README.md#attribute-limits)
to complex attributes.

> **Attribute value length limit** SHOULD be applied to all leaf string nodes in

Reviewer 7: ```suggestion
applied to all leaf string nodes
```
Should it also apply to `bytes` type?

Reviewer 2: 👍

how about:

```suggestion
**Attribute value length limit** SHOULD be applied to all leaf `string` and `bytes` nodes in
```

**Attribute value length limit** SHOULD be applied to all leaf string nodes in
`AnyValue` which SHOULD be truncated to the configured limit.

> Leaf nodes of an `AnyValue` attribute SHOULD count toward the **attribute count limit**.

Author: document that we should not add attribute when limit is reached

the SDK MUST not add the entire `AnyValue` attribute; partial exports are not allowed.

> [!NOTE]
>
> We should define `AnyValue` limits even if this OTEP is not merged -
> [#4487](https://github.com/open-telemetry/opentelemetry-specification/issues/4487)

### Exporters

> OTLP exporter SHOULD, by default, pass `AnyValue` attributes to the endpoint.

Reviewer 7: If it is "by default", what are non-default options? Does this imply configurability or just gives a leeway to implementations to do something else?

Reviewer: This statement also conflicts with

> Stable exporters will be prohibited from emitting complex attributes by default on signals other than Logs until at least Oct 1, 2025.

I guess the intention is that after Oct 1, 2025 all OTLP exporters should pass `AnyValue` attributes to the endpoint.

Before that date, by default, they should also emit complex attributes on LogRecords. Moreover, the OTLP exporters could have some opt-in configuration (e.g. env var) to allow complex attributes in other places.

Reviewer 2: > just gives a leeway to implementations to do something else?

@tigrannajaryan 👍

> This statement also conflicts with

@pellared I don't see the conflict, it just means you can't implement this OTEP in a stable exporter until at least Oct 1, 2025, can you explain a bit more?

Reviewer 7: Is there any way we can reword this? The word "default" to me means it is an adjustable runtime behavior. Just deleting "by default" is probably sufficient.

Reviewer 2: done in https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/85dd9b75247d757711cb55c4cf41fa4ae42c4eb0

SHOULD limit supported types to primitives, arrays, standard library collections,
named tuples, JSON objects, and similar structures following
[mapping to OTLP AnyValue](/specification/common/attribute-type-mapping.md#mapping-arbitrary-data-to-otlp-anyvalue).

Falling back to a string representation of unknown objects is RECOMMENDED to
minimize the risk of unintentional use of complex attributes.

Prior art on AnyValue conversion: [Go](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/72cccd8065dcfd84b69f34d8cb6f349a547fedce/bridges/otelslog/convert.go#L20),
[.NET](https://github.com/open-telemetry/opentelemetry-dotnet/blob/71abd4169b4b6c672343b37c32e3337bc227ed32/src/OpenTelemetry/Logs/ILogger/OpenTelemetryLogger.cs#L134),
[Python](https://github.com/open-telemetry/opentelemetry-python/blob/00329e07fb01d7c3e43bb513fe9be3748745c52e/opentelemetry-api/src/opentelemetry/attributes/__init__.py#L121)

## Prototypes

- [opentelemetry-python#4587](https://github.com/open-telemetry/opentelemetry-python/pull/4587)
- [opentelemetry-go#6809](https://github.com/open-telemetry/opentelemetry-go/pull/6180)

## Future possibilities

> ### Configurable OTLP exporter behavior

Reviewer 7: Is this about SDK exporter or Collector exporter?

Reviewer: Both :wink:

Reviewer 2: I think this section was originally intended for SDKs, but I agree it could apply to collector as well, updated in https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/b6390285dd421a7785500829ec961dab722bab0c

The OTLP exporter behavior for complex attributes can be made customizable on a per-signal
basis, allowing complex attributes to be:

- passed through (the default),
- serialized to JSON, or
- dropped

This option may be useful as a workaround for applications that don’t use a
collector and whose backend does not handle complex attribute types gracefully.

### Additional size limits

We can consider a separate set of attribute limits specifically for complex values,
including a total size limit and ability to estimate `AnyValue` object size.

## Backend research

See [the gist](https://gist.github.com/lmolkova/737ebba190b206a5d60bbc075fea538b)
for additional details.

| Backend                           | Handles complex attributes gracefully? | Comments        |

> | --------------------------------- | | ------------------------------ |

Reviewer 8: ```suggestion
| --------------------------------- | ----- | ------------------------------ |
```

Somehow the table is not formatting, this may help

| Jaeger (OTLP)                     | :white_check_mark: | serializes to JSON string |
| Prometheus with OTLP remote write | :white_check_mark: | serializes to JSON string |
| Grafana Tempo (OTLP)              | :white_check_mark: | serializes to JSON string, viewable but can't query using this attribute |
| Grafana Loki (OTLP)               | :white_check_mark: | flattens |
| Aspire dashboard (OTLP)           | :white_check_mark: | serializes to JSON string |
| ClickHouse (collector exporter)   | :white_check_mark: | serializes to JSON string, can parse JSON and query |

> | Honeycomb (OTLP)                  | :white_check_mark: | flattens if less than 5 layers deep, not array or binary data, JSON string otherwise |

Reviewer 9: ```suggestion
| Honeycomb (OTLP)                  | :white_check_mark: | flattens if less than 5 layers deep, not array or binary data, JSON string otherwise |
| Logfire (OTLP)                  | :white_check_mark: | stored as JSON, native support for JSON in queries |
```

> | New Relic (OTLP)                  | :white_check_mark: | drops the complex attribute | |

Reviewer: ```suggestion
| New Relic (OTLP)                  | :white_check_mark: | drops the complex attribute | |
| Splunk (OTLP)                     | :white_check_mark: | |
```

currently flattens for logs, serializes to JSON string for spans and metrics

  `bool`, `int64`, `double`, and `ArrayValue` of those types in the
  [AnyValue proto definition](https://github.com/open-telemetry/opentelemetry-proto/blob/42319f8b5bf330f7c3dd4a097384f9f6d5467450/opentelemetry/proto/common/v1/common.proto#L28-L40).
  These are currently referred to as *standard* attributes in the
  [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.44.0/specification/common/README.md)

- **Complex attributes** include all other values supported by the `AnyValue` proto,
  such as null (empty) value, maps, heterogeneous arrays, and combinations of those with primitives.
  Byte arrays are also considered complex attributes, as they are excluded from
  the current definition of *standard* attributes.

This distinction is not intended for the spec language, but is helpful here
because the OTEP proposes including both *simple* and *complex* attributes
in the set of *standard* attributes.

## Why?

### Why do we want complex attributes on spans?

There are a number of reasons why we want to allow complex attributes on spans:

- Emerging semantic conventions, such as Generative AI and GraphQL,

>   demonstrate the usefulness of including complex attributes on spans.

Reviewer 10: I think a citation or example could strengthen this point.

Reviewer 2: added in https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/cf9af5001c8c8ae86dccd22c5467d5dec9fdb815

- Many users already add complex attributes to spans by JSON-encoding them.
  Supporting complex attributes natively enables future possibilities like
  collector transformations, better attribute truncation, and native
  storage of complex attributes in backends.
- During the span event deprecation process, we found that some people
  use span events to represent complex data on spans.
  Providing support for complex attributes on spans would offer a more natural
  way to store this data.
- During the event stabilization process, it became clear that
  [spans and events are often thought of as being conceptual
  twins](https://opentelemetry.io/blog/2025/opentelemetry-logging-and-you/#how-is-this-different-from-other-signals),
  and that the choice between modeling something as a span or an event should not

>   be influenced by whether complex attributes are needed.

Reviewer 10: We could also clarify that the logs data model already supports `any` in attributes, which makes the signals inconsistent. The change in this otep makes spans and logs more consistent.

extending the standard attributes provides a more seamless and user-friendly API experience.

### Why doesn't this require a major version bump?

Currently, the SDK specification has a clause that says extending
the set of standard attribute would be
[considered a breaking change](/specification/common/README.md#standard-attribute).

We believe that removing this clause and extending standard
attributes can be done gracefully across the OpenTelemetry ecosystem
without requiring a major version bump:

- Language SDKs can implement this without breaking their backwards
  compatibility guarantees (e.g., [Java's](https://github.com/open-telemetry/opentelemetry-java/blob/main/VERSIONING.md)).
- While backends may still need to add support for complex attributes,
  this is the case with the introduction of any new OTLP feature.
  - Bumping the OTLP minor version is already the normal communication mechanism
    for this kind of change.
  - SDKs will be required to only emit complex attributes under that OTLP version
    or later.
  - Stable SDKs will be prohibited from emitting complex attributes until at

>     least Oct 1, 2025.

Reviewer 2: I've updated this line to

> Stable SDKs will be prohibited from emitting complex attributes on signals other than Logs until at least Oct 1, 2025.

  - A reasonably straightforward implementation option for backends is to just
    JSON serialize complex attributes and store them as strings.

## How

### API

Existing APIs that create or add attributes will be extended to support value
with type `AnyValue`.

It's RECOMMENDED to expose an `AnyValue` API for type checking, ergonomics,
and performance.

> Exposing multiple types of attribute sets is NOT RECOMMENDED, such as having "ExtendedAttributes" in addition "Attributes".

Reviewer 10: ```suggestion
Exposing multiple types of attribute sets is NOT RECOMMENDED, such as having "ExtendedAttributes" in addition to "Attributes".
```

OTel API MUST support setting complex attributes on spans, logs, profiles,
descriptive entity attributes.

OTel API SHOULD NOT provide convenience methods to set complex attributes on

> metrics, resources, instrumentation scope. APIs that allow to set complex attributes

Author: there is (no?) reason to not use complex attrs as descriptive on instrumentation scope or resource, so do we need to mention them here?

- While backends may still need to add support for complex attributes,
  this is the case with the introduction of any new OTLP feature.
  - Bumping the OTLP minor version is already the normal communication mechanism
    for this kind of change.
  - SDKs will be required to only emit complex attributes under that OTLP version
    or later.
  - Stable exporters will be prohibited from emitting complex attributes by default on signals
    other than Logs until at least Oct 1, 2025.
  - A reasonably straightforward implementation option for backends is to just
    JSON serialize complex attributes and store them as strings.

## How

### API

Existing APIs that create or add attributes will be extended to support value
with type `AnyValue`.

It's RECOMMENDED to expose an `AnyValue` API for type checking, ergonomics,
and performance.

Exposing multiple types of attribute sets is NOT RECOMMENDED, such as having "ExtendedAttributes" in addition to "Attributes".

OTel API MUST support setting complex attributes on spans, logs, profiles,
span links, exemplars, and as descriptive entity attributes.

> OTel API MAY support setting complex attributes on metrics, resources,

Reviewer 11: nit: It'd be helpful to add a note about why we picked MAY for metrics, scope , resources, and MUST for spans. (the MUST for Span is already covered by "### Why do we want complex attributes on spans?" section)

Reviewer 2: @cijothomas I added this note:

> [!NOTE]
> "MAY" is used here instead of "MUST" to give flexibility to dynamically typed language APIs since there are no concrete use cases at this time requiring complex attributes in these areas.
>
> Most likely statically typed languages will choose to support setting complex attributes uniformly everywhere.
>
> This requirement level could change from "MAY" to "MUST" in the future if we uncover use cases for complex attributes in these areas.

  this is the case with the introduction of any new OTLP feature.
  - Bumping the OTLP minor version is already the normal communication mechanism
    for this kind of change.
  - SDKs will be required to only emit complex attributes under that OTLP version
    or later.
  - Stable exporters will be prohibited from emitting complex attributes by default on signals
    other than Logs until at least 6 months after this OTEP is merged.
  - A reasonably straightforward implementation option for backends is to just
    JSON serialize complex attributes and store them as strings.

## How

### API

Existing APIs that create or add attributes will be extended to support
*complex attributes*.

It's RECOMMENDED to expose an `AnyValue` type - the API representing complex or
simple attribute value for type checking, ergonomics, and performance reasons.

Exposing multiple types of attribute sets is NOT RECOMMENDED, such as having "ExtendedAttributes" in addition to "Attributes".

OTel API MUST support setting complex attributes on spans, logs, profiles,
span links, and as descriptive entity attributes.

OTel API MAY support setting complex attributes on metrics, resources,

> instrumentation scope, span events, and as identifying entity attributes.

Author: Based on the last week Spec call, we still have some concerns on this line 

(along with "Based on the last week Spec call (7/1), we still have some concerns on this line Exposing multiple types of attribute sets is NOT RECOMMENDED, such as having "ExtendedAttributes" in addition to "Attributes".).

@dyladan please correct me if I'm wrong, but the summary is that:
- we don't have a real use-case for complex attributes on metrics or entities 
- it feels wrong to allow it and immediately tell users not to use it.

-----

I wanted to discuss a couple of points on the spec call tomorrow:
 
While we don't have a use-case in semantic conventions or instrumentations, it does not mean that end users don't have it or won't have it.

There are at least a few hypothetical use-cases where complex and structured data naturally appears out there in the wild:
    - Cloud resource information such as [VM metadata](https://learn.microsoft.com/azure/virtual-machines/instance-metadata-service?tabs=linux)
    - [container image manifests](https://jfrog.com/blog/the-more-the-merrier-multi-arch-docker-manifests-with-buildx-and-artifactory/)
    -  identity - some of service principal info can easily appear on entities / metrics  and can be [fairly complex](https://learn.microsoft.com/en-us/graph/api/resources/serviceprincipal?view=graph-rest-1.0#properties)
    - rarely occuring metrics where cardinality is not a issue - e.g. [GitHub PR](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request) info is full of nested objects

Complex data can appear on any signal and is discouraged in the same way on each of them. We're adding it to spans not because it's awesome, but because we need a way to record complex data that appears out there in the wild.

Complex data, especially if it's small, does not create much overhead over the same properties recorded as flat ones, but the bonus is that data on telemetry could match something well known defined outside of the telemetry.

language similar to this:

> Simple attributes SHOULD be used whenever possible. Instrumentations SHOULD
> assume that backends do not index individual properties of complex attributes,
> that querying or aggregating on such properties is inefficient and complicated,
> and that reporting complex attributes carries higher performance overhead.

### SDK

OTel SDK MUST support setting complex attributes on spans, logs, profiles,
span links, and as descriptive entity attributes.

OTel SDK MAY support setting complex attributes on metrics, exemplars, resources,
instrumentation scope, span events, and as identifying entity attributes.

> [!NOTE]
> "MAY" is used here instead of "MUST" to give flexibility to dynamically
> typed language SDKs since there are no concrete use cases at this time
> requiring complex attributes in these areas.
>
> Most likely statically typed languages will choose to support
> setting complex attributes uniformly everywhere.
>
> This requirement level could change from "MAY" to "MUST" in the future
> if we uncover use cases for complex attributes in these areas.

The SDK MUST support reading and modifying complex attributes during processing.

> #### `AnyValue` implementation notes

Author: let's make it conditional - if you allow complex attributes on entities/metrics, this is SHOULD, otherwise, doesn't apply

<!-- toc -->

- [Glossary](#glossary)
- [Why?](#why)
  * [Why do we want complex attributes on spans?](#why-do-we-want-complex-attributes-on-spans)
  * [Why do we want to extend standard attributes?](#why-do-we-want-to-extend-standard-attributes)
  * [Why doesn't this require a major version bump?](#why-doesnt-this-require-a-major-version-bump)
- [How](#how)
  * [API](#api)
  * [SDK](#sdk)
    + [`AnyValue` implementation notes](#anyvalue-implementation-notes)
    + [Attribute limits](#attribute-limits)
  * [Exporters](#exporters)
  * [Semantic conventions](#semantic-conventions)
  * [Proto](#proto)
- [Trade-offs and mitigations](#trade-offs-and-mitigations)
  * [Backends don't support `AnyValue` attributes](#backends-dont-support-anyvalue-attributes)
  * [Arbitrary objects are dangerous](#arbitrary-objects-are-dangerous)
- [Prototypes](#prototypes)
- [Future possibilities](#future-possibilities)
  * [Configurable OTLP exporter behavior (both SDK and Collector)](#configurable-otlp-exporter-behavior-both-sdk-and-collector)
  * [Additional size limits](#additional-size-limits)
- [Backend research](#backend-research)

<!-- tocstop -->

> ## Glossary

Author: Added 

> - **AnyValue** represents the type of *any* (simple or complex) attribute value on 
  the API, SDK, and [proto level](https://github.com/open-telemetry/opentelemetry-proto/blob/42319f8b5bf330f7c3dd4a097384f9f6d5467450/opentelemetry/proto/common/v1/common.proto#L28-L40).

>  It's also known as `any` in the [Log data model](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.44.0/specification/logs/data-model.md#type-any)

  These are currently referred to as *standard* attributes in the
  [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.44.0/specification/common/README.md)

- **Complex attributes** include all other values supported by the `AnyValue` proto,
  such as null (empty) value, maps, heterogeneous arrays, and combinations of those with primitives.
  Byte arrays are also considered complex attributes, as they are excluded from
  the current definition of *standard* attributes.

This distinction is not intended for the spec language, but is helpful here
because the OTEP proposes including both *simple* and *complex* attributes
in the set of *standard* attributes.

## Why?

### Why do we want complex attributes on spans?

There are a number of reasons why we want to allow complex attributes on spans:

- Emerging semantic conventions have demonstrated the usefulness of
  including complex attributes on spans, such as
  [capturing prompts and completions](https://github.com/open-telemetry/semantic-conventions/pull/2179)
  for Generative AI
  and [capturing request errors](https://graphql.org/learn/response/#request-errors)
  for GraphQL.

> - Many users already add complex attributes to spans by JSON-encoding them.

Reviewer 12: Not sure why this transformation couldn't already be done by adding a JSON transformation in collectors. ?

Author: it could be done, the point here is that if users do it already, we could provide support for complex attributes on the API + SDK level instead of asking them to serialize to json.

### SDK

OTel SDK MUST support setting complex attributes on spans, logs, profiles,
span links, and as descriptive entity attributes.

OTel SDK MAY support setting complex attributes on metrics, exemplars, resources,
instrumentation scope, span events, and as identifying entity attributes.

> [!NOTE]
> "MAY" is used here instead of "MUST" to give flexibility to dynamically
> typed language SDKs since there are no concrete use cases at this time
> requiring complex attributes in these areas.
>
> Most likely statically typed languages will choose to support
> setting complex attributes uniformly everywhere.
>
> This requirement level could change from "MAY" to "MUST" in the future
> if we uncover use cases for complex attributes in these areas.

The SDK MUST support reading and modifying complex attributes during processing.

#### `AnyValue` implementation notes

`AnyValue` implementation SHOULD provide deep equality check
since its possible (but not recommended) to use complex attributes as a resource,
instrumentation scope, or metric attribute where equality may be
extensively used to identify tracer/meter/logger or the time series.

> Equality of `KeyValueList`s MUST be unaffected by the ordering of their `KeyValue` pairs.

Reviewer 13: Note: this also means we need SDKs to deal with duplicate key issue.  Today we "allow" that to pass all the way downstream and deal with it later.

Probably not an issue but something to call out.

Reviewer: >  Today we "allow" that to pass all the way downstream and deal with it later.

I am not sure if this is true.

See:
1. proto: https://github.com/open-telemetry/opentelemetry-proto/blob/c30610041736aa5c0077b156f27b09e878b797ea/opentelemetry/proto/common/v1/common.proto#L57-L59

```
message KeyValueList {
  // A collection of key/value pairs of key-value pairs. The list may be empty (may
  // contain 0 elements).
  // The keys MUST be unique (it is not allowed to have more than one
  // value with the same key).
  repeated KeyValue values = 1;
}
```

2. logs data model: https://github.com/open-telemetry/opentelemetry-specification/blob/ac137b349b006dc926da91c6a6619217737921cc/specification/logs/data-model.md?plain=1#L129-L132

Author: since duplicate keys are not valid (but possible), I added 

> the equality behavior is unspecified if duplicate keys are present.

> Equality of `KeyValueList`s MUST be unaffected by the ordering of their `KeyValue` pairs.

Author: Updated to 

> When `AnyValue` contains one or more lists of key-value pairs:
> - the equality of `AnyValue` instances MUST NOT be affected by the ordering of 
>   the key-value pairs within the list.
> - the equality behavior is unspecified if duplicate keys are present.

Still working on the `AnyValue` definition

> #### Attribute limits

Reviewer 14: +1 to scoping discussion around attribute limits outside of this PR.

I think keeping the OTEP scoped to adding complex attributes and mentioning attribute limits as an "outstanding item" will help progress this PR and separate concerns around limits to their own targeted discussion.

> #### Attribute limits

Author: removed attribute limits from this OTEP in https://github.com/open-telemetry/opentelemetry-specification/pull/4485/commits/c1d9a5a684f5b611ae0750695a8727392989f053

The SDK SHOULD apply [attribute limits](/specification/common/README.md#attribute-limits)
to complex attributes.

> **Attribute value length limit** SHOULD be applied to all leaf `string` and `bytes` nodes in

Reviewer 13: I think the main change I'd like to see is *key name* length included in attribute length limit.

This way, for a highly complex attribute, you're "close" to the real size limit you want, even if serializing to JSON may bloat the size with `{`, `}`, `:` and `"`.

> **Attribute value length limit** SHOULD be applied to all leaf `string` and `bytes` nodes in

Author: @jsuereth since complex attributes are already part of the spec and don't have reasonable limits defined, we'll need to revisit their limits regardless of the outcome of this OTEP, I just removed the corresponding section from this OTEP in favor of https://github.com/open-telemetry/opentelemetry-specification/issues/4487. 

LMK if that's ok with you

> **Attribute value length limit** SHOULD be applied to all leaf `string` and `bytes` nodes in

Reviewer 13: sounds good to me!

> **Attribute value length limit** SHOULD be applied to all leaf `string` and `bytes` nodes in

Reviewer 13: I will see I think this OTEP *dramatically* incresaes priority of fixing this, hence my comments.  We can address on existing issues/PRs.

> `AnyValue` which SHOULD be truncated to the configured limit.

Author: mentioned this one in #4487 so we don't forget to address it

lexicographically and apply additional settings that enhance serialization stability.

### Semantic conventions

Semantic conventions will be updated with the following guidance:

- Simple attributes SHOULD be used whenever possible. Semantic conventions SHOULD
  assume that backends do not index individual properties of complex attributes,
  that querying or aggregating on such properties is inefficient and complicated,
  and that reporting complex attributes carries higher performance overhead.

- Complex attributes that are likely to be large (when serialized to a string) SHOULD
  be added as *opt-in* attributes. *Large* will be defined based on common backend
  attribute limits, typically around 4–16 KB.

- Complex attributes MUST NOT be used on metrics, resources, instrumentation scopes,
  or as identifying attributes on entities.

### Proto

OTLP uses `AnyValue` attributes on all signals, so the changes would be limited
to updating comments like [this one](https://github.com/open-telemetry/opentelemetry-proto/blob/be5d58470429d0255ffdd49491f0815a3a63d6ef/opentelemetry/proto/trace/v1/trace.proto#L209-L213)
and adding changelog record.

> ## Trade-offs and mitigations

Reviewer 12: What seems to be lacking is a considered alternative ?

Author: @hdost could you elaborate? what's the alternative?

Reviewer 12: The point here is that the solution seems to be leading the problem here. I am sure that the intentions are good, but I'd argue that adding this first class support for these complex attributes risks increasing the burden of comprehensability as well as introducing inefficiencies unnecessarily to the 99% case. 
We're now making attributes not as universal creating caveats to when you should use certain types of attributes.
My alternative would be don't add this, and if desired add extensions to improve their support without altering the standard interface.
The prescient thing here is that there is no mentioned alternative meaning to any reader that there really hasn't been any alternatives considered.

Author: @hdost the motivation is mentioned in the [Why](https://github.com/open-telemetry/opentelemetry-specification/blob/c1d9a5a684f5b611ae0750695a8727392989f053/oteps/4485-extending-attributes-to-support-complex-values.md#why) section. 

The data we want to express is sometimes complex and flattening is not an option. Complex attributes are already in the spec and are allowed on the logs. Users already put JSON blobs on the attributes serialized in all possible (sometimes invalid) ways. 

If you can propose a reasonable alternative that *allows to express complex data*, I'd be happy to mention it here and consider changing the direction.

Reviewer 12: So to your point Logs already carry and allow for this burden. Logs already have associated trace identifiers, so why then is it not sufficient to have users Log with these attributes? What is the advantage vs using logging?

Author: the need to record a log only to record complex attribute results in a poor user experience for everyone - instrumentation authors and end users. It also bloats telemetry size

It's in the why section - 

https://github.com/open-telemetry/opentelemetry-specification/blob/c1d9a5a684f5b611ae0750695a8727392989f053/oteps/4485-extending-attributes-to-support-complex-values.md?plain=1#L55-L74

### Backends don't support `AnyValue` attributes

While it should be possible to record complex data in telemetry, many backends do not
support it, which can result in individual complex attributes being dropped.

We mitigate this through:

- Introducing new APIs in the experimental parts of the OTel API which will limit
  the impact of unsupported attribute types to early adopters, while giving
  backends time to add support.

- Semantic conventions guidance that limits usage of complex attributes.

- Existing collector transformation processor that can drop, flatten, serialize,
  or truncate complex attributes using OTTL.

### Arbitrary objects are dangerous

Allowing arbitrary objects as attributes is convenient but increases the risk of
including large, sensitive, mutable, non-serializable, or otherwise problematic
data in telemetry.

> It is RECOMMENDED to enforce `AnyValue` compatibility at the API level.

Author: removed this sentence since it the recommendation would be too vague and leave it up to the implementation SIG choice anyway.

The intent was to recommend having explicit AnyValue APIs for statically typed languages.

-----

E.g. in Java/.NET the recommendation would to be to do something like

```java
// or maybe I can extend AnyValue type and have my custom FooBarValue
var myFooObj = AnyValue.builder()
   .addStringValue("foo", "value")
   .addIntValue("bar", 42)
   .build();

var attributes = Attributes.builder()
  .put(anyValueKey("foo", myFooObj))  // takes AnyValue as a value
  .build();
```

You can also have an API that takes arbitrary interface/object/etc and converts it to proto `AnyValue` under the hood. 

```pseudo-java

object myFooObj = getArbitraryObject();
var attributes = Attributes.builder()
  .put(anyValueKey("foo", myFooObj)) // takes arbitrary object as a value
  .build();
```

The latter one is convenient, but dangerous and not performant. While its ok to provide it with some limitations, it's likely a convenience on top of the former 'explicit AnyValue` API and is not a replacement for it.
